lib/vfs/src/filesystem_ext.rs

```diff
@@ -0,0 +1,238 @@
+use std::{collections::VecDeque, path::Path};
+
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+use crate::{DirEntry, FileSystem, FsError};
+
+/// Helper methods for working with [`FileSystem`]s.
+#[async_trait::async_trait]
+pub trait FileSystemExt {
+    /// Does this item exists?
+    fn exists(&self, path: impl AsRef<Path>) -> bool;
+
+    /// Does this path refer to a directory?
+    fn is_dir(&self, path: impl AsRef<Path>) -> bool;
+
+    /// Does this path refer to a file?
+    fn is_file(&self, path: impl AsRef<Path>) -> bool;
+
+    /// Make sure a directory (and all its parents) exist.
+    ///
+    /// This is analogous to [`std::fs::create_dir_all()`].
+    fn create_dir_all(&self, path: impl AsRef<Path>) -> Result<(), FsError>;
+
+    /// Asynchronously write some bytes to a file.
+    ///
+    /// This is analogous to [`std::fs::write()`].
+    async fn write(
+        &self,
+        path: impl AsRef<Path> + Send,
+        data: impl AsRef<[u8]> + Send,
+    ) -> Result<(), FsError>;
+
+    /// Asynchronously read a file's contents into memory.
+    ///
+    /// This is analogous to [`std::fs::read()`].
+    async fn read(&self, path: impl AsRef<Path> + Send) -> Result<Vec<u8>, FsError>;
+
+    /// Asynchronously read a file's contents into memory as a string.
+    ///
+    /// This is analogous to [`std::fs::read_to_string()`].
+    async fn read_to_string(&self, path: impl AsRef<Path> + Send) -> Result<String, FsError>;
+
+    /// Update a file's modification and access times, creating the file if it
+    /// doesn't already exist.
+    fn touch(&self, path: impl AsRef<Path> + Send) -> Result<(), FsError>;
+
+    /// Recursively iterate over all paths inside a directory, ignoring any
+    /// errors that may occur along the way.
+    fn walk(&self, path: impl AsRef<Path>) -> Box<dyn Iterator<Item = DirEntry> + '_>;
+}
+
+#[async_trait::async_trait]
+impl<F: FileSystem + ?Sized> FileSystemExt for F {
+    fn exists(&self, path: impl AsRef<Path>) -> bool {
+        self.metadata(path.as_ref()).is_ok()
+    }
+
+    fn is_dir(&self, path: impl AsRef<Path>) -> bool {
+        match self.metadata(path.as_ref()) {
+            Ok(meta) => meta.is_dir(),
+            Err(_) => false,
+        }
+    }
+
+    fn is_file(&self, path: impl AsRef<Path>) -> bool {
+        match self.metadata(path.as_ref()) {
+            Ok(meta) => meta.is_file(),
+            Err(_) => false,
+        }
+    }
+
+    fn create_dir_all(&self, path: impl AsRef<Path>) -> Result<(), FsError> {
```

Comment:
That's fine with me. 

The main reason I introduced an extension trait in the first place was that doing anything with a `FileSystem` would require several lines of boilerplate that you'd copy/paste every time. That was especially annoying in tests where using 4 lines when 1 is necessary can make them hard to follow.

I don't mind if we put them all on the main trait instead of using an extension trait.

Reply:
Move this as a standalone function in the crate instead of the Trait

Reply:
I've moved the functions to a module called `ops`.

Not being able to use method syntax takes away a lot of the ergonomics gains the extension trait gave us, so I'm not going to expose the functions as part of the public API.